### ISSUE TITLE
Unable to update argo appProject in UI without extra permissions

### ISSUE THREAD

**Issue by ragarcia26** [argoproj/argo-cd#3701]
**Describe the bug**

After creating a new appProject via the UI, I'm unable to update the Repo source list or the Cluster whitelist, even though I have full permissions over the project. I get the error:
`Unable to edit project: permission denied: repositories, update, https://MySecondRepository.git`  
or  
`Unable to edit project: permission denied: clusters, update, https://ClusterURL.com`  
I'm wondering why I need update permissions for the new repo/cluster I'm trying to add to my project. I can add it when first creating my project, but not when updating it.

**To Reproduce**

Relevant RBAC:
```
    p, role:test, clusters, get, *, allow
    p, role:test, repositories, get, *, allow
    p, role:test, projects, *, test, allow
```
1. Using the above role create a new project with a repo and cluster
2. Edit project and add a new repo and cluster

**Expected behavior**

Project updates with new cluster/repo.

**Argocd Version**

```shell
v1.6.0+43d2184
```

**Comment by alexmt** [comment]
Hello @ragarcia26 ,

Looked at the project update API implementation and discovered that check is intentional: https://github.com/argoproj/argo-cd/issues/1432

I'm not 100% if the implementation is correct. Either way, we need to document which permissions the user should have to update the project. Will post update when I find if implementation is correct or not and document the behavior.

**Comment by ragarcia26** [comment]
ok well thats an interesting change. I disagree that anyone adding a cluster or repo to whitelist permissions makes them admin over said cluster/repo, it just means they can use those resources, and you can already can limit what cluster/repo those users can use by rbac get. Which is what i was trying to do. It's fine, Users should be responsible enough to not blow up configurations owned by them. Thanks for the insight.

**Comment by alexmt** [comment]
I think it makes sense to keep the ticket open and use it to make sure the implementation is correct. To be honest I have the same expectations as you @ragarcia26 . It feels like user should have get permissions, not update.

**Comment by ragarcia26** [comment]
Perhaps this derives from the default role being read all for any non authorized user. In this way, you'd have to disable that default read only role (what we had done), or use rbac to block get's to undesired clusters which may become tedious.

**Comment by alexmt** [comment]
@ragarcia26 this is exactly the reason. We could not use `get` action since it is likely that Argo CD is configured so that everyone has get access. So the choose was between `update` or some new action (e.g. `use`). Sound like this not a bug, the documentation is still required.

**Comment by agaudreault** [comment]
Some ideas I have on this:

1. If a user has the permission to create a project, do not check for cluster and repositories update. The create action has an upsert flag that allows to bypass these permissions anyway. (will not fix the issues if user cannot create project)
2. Introduce new patterns such as `p, user, projects, update/clusters, *, allow`,  `p, user, projects, update/repositories, *, allow` and  `p, user, projects, update/*, *, allow`. This way it does not give permissions over the repositories and clusters resources, but allows to restrict the scope of the project update. This would be a "breaking" change that would need to be in the release notes.
    3. Update the create with upsert and apply the same rbac restrictions than update

In my opinion, the behavior currently forces the user to give cluster and repositories updates permissions, giving them permission on cluster and repositories API that should not be necessary. And these permission can easily be bypassed using the Create project + upsert flag.

**Comment by crenshaw-dev** [comment]
I agree @agaudreault, I think we should just drop the restrictions that were added in https://github.com/argoproj/argo-cd/issues/1432. 

We don't do this for other resources: users are meant to understand the kind of access that one resource implies. Doing cross-resource checks complicates the mental model and the code.
